src/js/utils/fullscreen-api.js

```diff
@@ -1,27 +1,179 @@
+import { WebkitPresentationModes } from '../constants.js';
+import { containsComposedNode } from './element-utils.js';
 import { document } from './server-safe-globals.js';
 
-export const fullscreenApi = {
-  enter: 'requestFullscreen',
-  exit: 'exitFullscreen',
-  rootEvents: ['fullscreenchange'],
-  mediaEvents: [],
-  element: 'fullscreenElement',
-  error: 'fullscreenerror',
-  enabled: 'fullscreenEnabled',
+// NOTE: (re)defining these types, but more narrowly for API expectations. These should probably be centralized + derived
+// once migrated to TypeScript types (CJP)
+
+/**
+ * @typedef {Partial<HTMLVideoElement> & {
+ *  webkitDisplayingFullscreen?: boolean;
+ *  webkitPresentationMode?: 'fullscreen'|'picture-in-picture';
+ *  webkitEnterFullscreen?: () => any;
+ * }} MediaStateOwner
+ */
+
+/**
+ * @typedef {Partial<Document|ShadowRoot>} RootNodeStateOwner
+ */
+
+/**
+ * @typedef {Partial<HTMLElement>} FullScreenElementStateOwner
+ */
+
+/**
+ * @typedef {object} StateOwners
+ * @property {MediaStateOwner} [media]
+ * @property {RootNodeStateOwner} [documentElement]
+ * @property {FullScreenElementStateOwner} [fullscreenElement]
+ */
+
+/** @type {(stateOwners: StateOwners) => Promise<undefined> | undefined} */
+export const enterFullscreen = (stateOwners) => {
+  const { media, fullscreenElement } = stateOwners;
+
+  // NOTE: Since the fullscreenElement can change and may be a web component,
+  // we should not define this at the module level. As an optimization,
+  // we could only define/update this somehow based on state owner changes. (CJP)
+  const enterFullscreenKey =
+    'requestFullscreen' in fullscreenElement
+      ? 'requestFullscreen'
+      : 'webkitRequestFullScreen' in fullscreenElement
+      ? 'webkitRequestFullScreen'
+      : undefined;
+
+  // Entering fullscreen cases (browser-specific)
+  if (enterFullscreenKey) {
+    // NOTE: Since the "official" enter fullscreen method yields a Promise that rejects
+    // if already in fullscreen, this accounts for those cases.
+    const maybePromise = fullscreenElement[enterFullscreenKey]?.();
+    if (maybePromise instanceof Promise) {
+      return maybePromise.catch(() => {});
+    }
+  } else if (media?.webkitEnterFullscreen) {
+    // Media element fullscreen using iOS API
+    media.webkitEnterFullscreen();
+  } else if (media?.requestFullscreen) {
+    // So media els don't have to implement multiple APIs.
+    media.requestFullscreen();
+  }
+};
+
+const exitFullscreenKey =
```

Comment:
Per out of band convo, should these "prefer" browser prefix versions or standard versions?

Reply:
like this seems good to me, prefer prefix-less

Reply:
cool would love to get @heff's and/or @mmcc's gut on this as well

Reply:
Yeah, can't say I'm thinking too deeply on this one, but I also prefer prefixless, in the hopes we can just delete code later without changing much more.

Reply:
Yeah for me I just wasn't sure if there was an expectation that, if both are around, which would be more reliable. That said, I think the status quo wins by vote, so I'm gonna mash that merge.